x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java

```diff
@@ -410,49 +410,59 @@ private void processDeploymentStats(GetDeploymentStatsAction.Response statsRespo
-                ClientHelper.executeAsyncWithOrigin(
-                    client,
-                    ClientHelper.ML_ORIGIN,
-                    UpdateTrainedModelDeploymentAction.INSTANCE,
-                    updateRequest,
-                    ActionListener.wrap(updateResponse -> {
-                        logger.info("adaptive allocations scaler: scaled [{}] to [{}] allocations.", deploymentId, newNumberOfAllocations);
-                        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME)
-                            .execute(
-                                () -> inferenceAuditor.info(
-                                    deploymentId,
-                                    Strings.format(
-                                        "adaptive allocations scaler: scaled [%s] to [%s] allocations.",
-                                        deploymentId,
-                                        newNumberOfAllocations
-                                    )
-                                )
-                            );
-                    }, e -> {
-                        logger.atLevel(Level.WARN)
-                            .withThrowable(e)
-                            .log(
-                                "adaptive allocations scaler: scaling [{}] to [{}] allocations failed.",
-                                deploymentId,
-                                newNumberOfAllocations
-                            );
-                        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME)
-                            .execute(
-                                () -> inferenceAuditor.warning(
-                                    deploymentId,
-                                    Strings.format(
-                                        "adaptive allocations scaler: scaling [%s] to [%s] allocations failed.",
-                                        deploymentId,
-                                        newNumberOfAllocations
-                                    )
-                                )
-                            );
-                    })
-                );
+                updateNumberOfAllocations(deploymentId, newNumberOfAllocations);
             }
         }
     }
+
+    public boolean maybeStartAllocation(TrainedModelAssignment assignment) {
+        if (assignment.getAdaptiveAllocationsSettings() != null
+            && assignment.getAdaptiveAllocationsSettings().getEnabled() == Boolean.TRUE) {
+            lastScaleUpTimesMillis.put(assignment.getDeploymentId(), System.currentTimeMillis());
+            updateNumberOfAllocations(assignment.getDeploymentId(), 1);
+            return true;
+        }
+        return false;
+    }
+
+    private void updateNumberOfAllocations(String deploymentId, int numberOfAllocations) {
+        UpdateTrainedModelDeploymentAction.Request updateRequest = new UpdateTrainedModelDeploymentAction.Request(deploymentId);
```

Comment:
This action writes a control message to the pytorch_inference process that is handled here

https://github.com/elastic/ml-cpp/blob/main/bin/pytorch_inference/Main.cc#L130

There is logic in the thread pool that prevents this value going lower than 1

https://github.com/elastic/ml-cpp/blob/main/lib/core/CStaticThreadPool.cc#L55

Reply:
On the Java side, the process is terminated when you set it to zero allocations.

Here are the logs "proving" this:
```
[2024-09-25T12:39:08,781][DEBUG][o.e.x.m.i.a.AdaptiveAllocationsScaler] [runTask-0] [my-elser-model] adaptive allocations scaler: scaling down to zero, because of no requests.
[2024-09-25T12:39:08,782][DEBUG][o.e.x.m.i.a.AdaptiveAllocationsScaler] [runTask-0] [my-elser-model] adaptive allocations scaler: load in [0.000, 0.019], scaling from 1 to 0 allocations.
[2024-09-25T12:39:08,786][DEBUG][o.e.x.m.i.a.p.AllocationReducer] [runTask-0] [my-elser-model] removing assignment with [1] allocations on node [T5jZ_EaBQ3maMJ_jbgMJMA]
[2024-09-25T12:39:08,787][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [runTask-0] updated assignments: {"my-elser-model":{"task_parameters":{"model_id":".elser_model_2","deployment_id":"my-elser-model","model_bytes":438123914,"threads_per_allocation":1,"number_of_allocations":0,"queue_capacity":1024,"cache_size":"438123914b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0},"routing_table":{},"assignment_state":"started","start_time":"2024-09-24T08:57:27.976985Z","max_assigned_allocations":1,"adaptive_allocations":{"enabled":true}}}
[2024-09-25T12:39:08,842][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentNodeService] [runTask-0] [my-elser-model] Stopping unreferenced deployment for node T5jZ_EaBQ3maMJ_jbgMJMA
[2024-09-25T12:39:08,842][DEBUG][o.e.x.m.i.d.TrainedModelDeploymentTask] [runTask-0] [my-elser-model] Stopping due to reason [node no longer referenced in model routing table]
[2024-09-25T12:39:08,843][INFO ][o.e.x.m.i.d.DeploymentManager] [runTask-0] [my-elser-model] Stopping deployment, reason [node no longer referenced in model routing table]
[2024-09-25T12:39:08,843][DEBUG][o.e.x.m.i.d.DeploymentManager] [runTask-0] [my-elser-model] Forcefully stopping process
[2024-09-25T12:39:08,844][INFO ][o.e.x.m.i.a.AdaptiveAllocationsScalerService] [runTask-0] adaptive allocations scaler: scaled [my-elser-model] to [0] allocations.
[2024-09-25T12:39:08,885][DEBUG][o.e.x.m.i.p.p.PyTorchResultProcessor] [runTask-0] [my-elser-model] Results processing finished
[2024-09-25T12:39:08,932][INFO ][o.e.x.m.p.l.CppLogMessageHandler] [runTask-0] [controller/92137] [CDetachedProcessSpawner.cc@187] Child process with PID 92160 was terminated by signal 15

Reply:
So I don't think that C++ code is executed in this case. Or it's executed and right after the process is terminated. (If I understand correctly how all of this works.)